x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/CertParsingUtils.java

```diff
@@ -275,4 +276,20 @@ public static X509ExtendedTrustManager trustManager(KeyStore keyStore, String al
         }
         throw new IllegalStateException("failed to find a X509ExtendedTrustManager");
     }
+
+    /**
+     * Checks that the {@code X509Certificate} array is ordered, such that the endpoint's is first and it is followed by any certificate
+     * authorities'. The check validates that the {@code issuer} of every certificate is the {@code subject} of the certificate in the next
+     * array position. No other certificate attributes are checked.
```

Comment:
I dug into the cert path validation code and RFCs. I think we should **not** assert end entity quality on the first certificate.

`sun.security.provider.certpath.ConstraintsChecker` validates the CA attribute for all certs in the chain, but it does **not** check the first one (Same for `KeyChecker`, validating the key usage attribute). In [RFC 5280](tools.ietf.org/html/rfc5280) end entity certificate is loosely described as `"is a certificate that either does not contain a Basic Constraints extension, or that contains such an extension with the CA flag equal to FALSE"` but in another section it is said explicitly that `"Usually, the last certificate is an end entity certificate, but it can be a CA"` 

There is also the practical reason that a self signed certificate from the truststore (be it CA or not) should be "trusted", ie the client can use that and be authenticated.

Reply:
I think we need to decide what our contract is here.
We are trusting that the proxy (Kibana) performed a valid TLS handshake using that certificate.

Do we also want to duplicate the checks that the proxy ought to have performed? (The ones we can check, since we can't check that the end-client holds the proper key).  Maybe we do, but I don't think that's necessarily the case. We could reasonable decide either way.

If node.js understands & validates an extension that the JVM/JSSE doesn't, should that be an error?

The certificate ordering check is reasonable because would extract an incorrect DN/principal if the certs aren't in the right order, but we need to determine whether we want to take on full X.509 validation, given that we already know we can't validate the handshake.

Reply:
There is a distinction to make here.
On one hand there is the question of whether to assert the **lack** of the BasicConstraint, which confers the CA role and it is described here [rfc5280#section-4.2.1.9](https://tools.ietf.org/html/rfc5280#section-4.2.1.9), and on the other hand is if we validate the key usage extension such as it can be used for TLS authentication.

On the first matter, I read the RFC that there is no validation required that the first cert be an end entity. The section that specifically describes this constraint (BasicConstraint) only, is [rfc5280#section-4.2.1.9](https://tools.ietf.org/html/rfc5280#section-4.2.1.9), and it says clearly `This extension MAY appear as a critical or non-critical extension in end entity certificates.` This is in accordance with the excerpts you pasted, because the negation of the constraint is explained in `If the basic constraints extension is not present in a version 3 certificate, or the extension is present but the cA boolean is not asserted, then the certified public key MUST NOT be used to verify certificate signatures.` In a different section [section-6.1](https://tools.ietf.org/html/rfc5280#section-6.1) on the path validation algorithm it is mentioned `In most cases, the target certificate will be an end entity certificate, but the target   certificate may be a CA certificate as long as the subject public key is to be used for a purpose other than verifying the signature on a public key certificate.` .
For me, this means the relationship goes in the opposite direction, if you are a CA you must have this constraint, and not if you have this constraint you must be a CA.

On the second matter, I believe you are right we should be validating the key usage extension, but I haven't looked into that. I was hoping this was already handled? I will be looking into that.

Reply:
@bizybot in `sun.security.validator.Validator#validate` there is
```
        // omit EE extension check if EE cert is also trust anchor
        if (chain.length > 1) {
            // EndEntityChecker does not need to check unresolved critical
            // extensions when validating with a TYPE_PKIX Validator.
            // A TYPE_PKIX Validator will already have run checks on all
            // certs' extensions, including checks by any PKIXCertPathCheckers
            // included in the PKIXParameters, so the extra checks would be
            // redundant.
            boolean checkUnresolvedCritExts =
                    (type == TYPE_PKIX) ? false : true;
            endEntityChecker.check(chain[0], parameter,
                                   checkUnresolvedCritExts);
        }
```
This is where the key usage constraint is checked on the end entity certificate (if it is not a trust anchor).
So the check you are suggesting we perform is already done as part of the chain validation.
_I tried testing it but my openssl foo is weak._

Do you conclude the test is already handled in the validation we already perform and so we don't need to make any extra checks?